### PR TITLE
Send emails

### PR DESCRIPTION
**Pull request by marialivia16**
- Create email client
- Send emails when:
    - a commercial request has been sent
   - a new key has been created

### CODE REVIEW

File: app/email/AwsEmailClient.scala [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+      s"Product name: ${user.productName}\n" +
+      s"Product URL: ${user.productUrl}\n" +
+      s"Company name: ${user.companyName}\n" +
+      s"Company URL: ${user.companyUrl.getOrElse('-')}\n" +
+      s"Business area: ${user.additionalInfo.businessArea.getOrElse('-')}\n" +
+      s"Commercial model: ${user.additionalInfo.commercialModel.getOrElse('-')}\n" +
+      s"Content type: ${user.additionalInfo.content.getOrElse('-')}\n" +
+      s"Monthly users: ${user.additionalInfo.monthlyUsers.getOrElse('-')}\n" +
+      s"Articles per day: ${user.additionalInfo.articlesPerDay.getOrElse('-')}\n" +
+      s"http://localhost:9000/user/${user.bonoboId}/edit"
```
Comment (mchv) [inline app/email/AwsEmailClient.scala]: You can use this syntax:
```scala
controllers.Applicatio.editUser(user.bonoboId)
```

Comment (mchv) [inline app/email/AwsEmailClient.scala]: you can use a play template to fill the message

File: app/components/AppComponents.scala [hunk 2]
```diff
@@ -59,14 +61,27 @@ trait KongComponentImpl extends KongComponent { self: BuiltInComponents with Nin
...
-trait ControllersComponent { self: BuiltInComponents with NingWSComponents with GoogleAuthComponent with DynamoComponent with KongComponent =>
+trait AwsEmailComponent {
+  def awsEmail: MailClient
+}
+
+trait AwsEmailComponentImpl extends AwsEmailComponent { self: BuiltInComponents =>
+  val awsEmail = {
+    val awsRegion = Regions.fromName(configuration.getString("aws.region") getOrElse "eu-west-1")
+    val amazonSesClient: AmazonSimpleEmailServiceAsyncClient = new AmazonSimpleEmailServiceAsyncClient(CredentialsProvider.getCredentials).withRegion(awsRegion)
+    val fromAddress = "test@open-platform.theguardian.com"
```
Comment (mchv) [inline app/components/AppComponents.scala]: Adding a comment about the necessity to verify the address [with a link to the validator](http://docs.aws.amazon.com/ses/latest/DeveloperGuide/verify-email-addresses.html)

File: app/email/AwsEmailClient.scala [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+      Product name: ${user.productName}
+      Product URL: ${user.productUrl}
+      Company name: ${user.companyName}
+      Company URL: ${user.companyUrl.getOrElse('-')}
+      Business area: ${user.additionalInfo.businessArea.getOrElse('-')}
+      Commercial model: ${user.additionalInfo.commercialModel.getOrElse('-')}
+      Content type: ${user.additionalInfo.content.getOrElse('-')}
+      Monthly users: ${user.additionalInfo.monthlyUsers.getOrElse('-')}
+      Articles per day: ${user.additionalInfo.articlesPerDay.getOrElse('-')}
+      ${controllers.routes.Application.editUserPage(user.bonoboId).absoluteURL()}"""
```
Comment (cb372) [inline app/email/AwsEmailClient.scala]: Sorry I forgot to mention it, but you have to be careful with whitespace in multiline (`"""`) strings. In its current state, the email will look like:

```
Name: Chris
      Email: ...
      Product name: ...
```

You can fix this by either removing the whitespace in the source code, e.g.

```
    val message = s"""Name: ${user.name}
Email: ${user.email}
Product name: ...
..."""
```

or by using the `stripMargins` method:

```
    val message = s"""Name: ${user.name}
                  |Email: ${user.email}
                  |Product name: ...
                  |...""".stripMargins

File: conf/routes [hunk 4]
```diff
@@ -14,7 +14,7 @@ POST        /key/:id/edit              controllers.Application.editKey(id: Strin
 
 GET         /user/create               controllers.Application.createUserPage
 POST        /user/create               controllers.Application.createUser
-GET         /user/:id/edit             controllers.Application.editUserPage(id: String)
+GET         /user/:id/edit             controllers.Application.editUserPage(id: String, error: Option[String] ?= None)
```
Comment (cb372) [inline conf/routes]: Let's use flash scope instead of sending the error message as a query parameter

File: app/controllers/CommercialForm.scala [hunk 5]
```diff
@@ -21,15 +23,21 @@ class CommercialForm(dynamo: DB, kong: Kong, val messagesApi: MessagesApi) exten
...
+    def handleValidForm(formData: CommercialRequestKeyFormData): Future[Result] = {
       logic.sendRequest(formData) match {
-        case Left(error) => BadRequest(views.html.commercialRequestKey(requestKeyForm.fill(formData), error = Some(error)))
-        case Right(_) => Redirect(routes.CommercialForm.requestMessage())
+        case Left(error) => Future.successful(BadRequest(views.html.commercialRequestKey(requestKeyForm.fill(formData), error = Some(error))))
+        case Right(user) => {
+          awsEmail.sendEmailCommercialRequest(user) map {
+            result => Redirect(routes.CommercialForm.requestMessage())
+          } recover {
+            case _ => Redirect(routes.CommercialForm.requestMessage()).flashing("error" -> "We were unable to send the email. Please contact [email] for further instructions")
```
Comment (cb372) [inline app/controllers/CommercialForm.scala]: Add a TODO comment here to fix once we know the correct email address

File: app/email/AwsEmailClient.scala [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+      |Product URL: ${user.productUrl}
+      |Company name: ${user.companyName}
+      |Company URL: ${user.companyUrl.getOrElse('-')}
+      |Business area: ${user.additionalInfo.businessArea.getOrElse('-')}
+      |Commercial model: ${user.additionalInfo.commercialModel.getOrElse('-')}
+      |Content type: ${user.additionalInfo.content.getOrElse('-')}
+      |Monthly users: ${user.additionalInfo.monthlyUsers.getOrElse('-')}
+      |Articles per day: ${user.additionalInfo.articlesPerDay.getOrElse('-')}
+      |${controllers.routes.Application.editUserPage(user.bonoboId).absoluteURL()}""".stripMargin
+    sendEmail("maria-livia.chiorean@guardian.co.uk", "Commercial Key Request", message) //this should be eventually sent to content.delivery@guardian.co.uk instead
```
Comment (cb372) [inline app/email/AwsEmailClient.scala]: Please add the word "TODO" to this comment

File: app/email/AwsEmailClient.scala [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+  }
+
+  def sendEmailNewKey(toEmail: String, key: String): Future[SendEmailResult] = {
+    val message = s"A new key has been created for you: $key"
+    sendEmail(toEmail, "New Key Created", message)
+  }
+}
+
+object AwsMailClient {
```
Comment (cb372) [inline app/email/AwsEmailClient.scala]: I don't think this is used any more

File: app/views/renderMessage.scala.html [hunk 8]
```diff
@@ -7,4 +7,9 @@
     @for(e <- error) {
         <p class="alert alert-danger">@e</p>
     }
+    @flash.map { e =>
```
Comment (cb372) [inline app/views/renderMessage.scala.html]:
```
@flash.flatMap(_.get("error")) { e =>
  <p class="...">@e</p>
}
```

**Comment by cb372** [comment]
Please add a new policy to the instance profile in CloudFormation.

http://docs.aws.amazon.com/ses/latest/DeveloperGuide/control-user-access.html#iam-and-ses-examples-email-sending-actions

**Comment by cb372** [comment]
:+1: !!

**Comment by cb372** [comment]
We should update the CloudFormation stack to add the permissions before merging this.

**Pull request merged**